crates/ty_python_semantic/src/types/infer/builder/type_expression.rs

```diff
@@ -1280,10 +1280,27 @@ impl<'db> TypeInferenceBuilder<'db, '_> {
                         .top_materialization(self.db()),
                 ),
             },
-            SpecialFormType::TypeGuard => {
-                self.infer_type_expression(arguments_slice);
-                todo_type!("`TypeGuard[]` special form")
-            }
+            SpecialFormType::TypeGuard => match arguments_slice {
+                ast::Expr::Tuple(_) => {
+                    self.infer_type_expression(arguments_slice);
+
+                    if let Some(builder) = self.context.report_lint(&INVALID_TYPE_FORM, subscript) {
+                        let diag = builder.into_diagnostic(format_args!(
+                            "Special form `{}` expected exactly one type parameter",
+                            special_form.repr()
+                        ));
+                        diagnostic::add_type_expression_reference_link(diag);
+                    }
+
+                    Type::unknown()
+                }
+                _ => TypeGuardType::unbound(
+                    self.db(),
+                    // Similar to TypeIs, use top materialization
+                    self.infer_type_expression(arguments_slice)
+                        .top_materialization(self.db()),
```

Comment:
As mentioned in another comment, I think for `TypeGuard`, given its "clobber" semantics, we maybe shouldn't do this? Or at least I'm curious what the impact on the ecosystem report would be. We do see some false positives in the ecosystem from doing this.